src/libcore/slice/mod.rs

```diff
@@ -2405,8 +2422,21 @@ macro_rules! iterator {
 
             #[inline]
             fn nth(&mut self, n: usize) -> Option<$elem> {
-                // Call helper method. Can't put the definition here because mut versus const.
-                self.iter_nth(n)
+                if n >= self.len() {
+                    // This iterator is now empty.  The way we encode the length of a non-ZST
+                    // iterator, this works for both ZST and non-ZST.
+                    // For a ZST we would usually do `self.end = self.ptr`, but since
+                    // we will not give out an reference any more after this there is no
+                    // way to observe the difference except for raw pointers.
+                    self.ptr = self.end;
```

Comment:
Previously in `next` there's an `assume(!self.end.is_null())` but only if `T` has a nonzero size. That implies to me that if T is a ZST it's possible for `self.end` to be null (I think wraparound?). In this case, this is setting `self.ptr` to null, right? Wouldn't that break the in-memory representation of `Option<&[T]>` as well as the first `assume` above in the call to `next`?

Reply:
Oh good point, I don't think we can rule that out.

I guess I'll hope for the optimizer to remove the conditional here as well, then. :)